components/api/src/main/java/com/hotels/styx/api/messages/FullHttpRequest.java

```diff
@@ -0,0 +1,680 @@
+        private boolean validate = true;
+        private Url url;
+        private boolean secure;
+        private HttpHeaders.Builder headers;
+        private HttpVersion version = HTTP_1_1;
+        private T body;
+        private final List<HttpCookie> cookies;
+
+        public Builder() {
+            this.url = Url.Builder.url("/").build();
+            this.headers = new HttpHeaders.Builder();
+            this.body = null;
+            this.cookies = new ArrayList<>();
+        }
+
+        public Builder(HttpMethod method, String uri) {
+            this();
+            this.method = requireNonNull(method);
+            this.url = Url.Builder.url(uri).build();
+            this.secure = url.isSecure();
+        }
+
+        public Builder(HttpRequest request, T body) {
+            this.id = request.id();
+            this.method = request.method();
+            this.clientAddress = request.clientAddress();
+            this.url = request.url();
+            this.secure = request.isSecure();
+            this.version = request.version();
+            this.headers = request.headers().newBuilder();
+            this.body = body;
+            this.cookies = new ArrayList<>(request.cookies());
+        }
+
+        Builder(FullHttpRequest<T> request) {
+            this.id = request.id();
+            this.method = request.method();
+            this.clientAddress = request.clientAddress();
+            this.url = request.url();
+            this.secure = request.isSecure();
+            this.version = request.version();
+            this.headers = request.headers().newBuilder();
+            this.body = request.body();
+            this.cookies = new ArrayList<>(request.cookies());
+        }
+
+        /**
+         * Sets the request URI.
+         *
+         * @param uri URI
+         * @return {@code this}
+         */
+        public Builder<T> uri(String uri) {
+            return this.url(Url.Builder.url(uri).build());
+        }
+
+        /**
+         * Sets the request body.
+         *
+         * @param content request body
+         * @return {@code this}
+         */
+        public Builder<T> body(T content) {
+            this.body = content;
+            return this;
+        }
+
+        /**
+         * Sets the unique ID for this request.
+         *
+         * @param id request ID
+         * @return {@code this}
+         */
+        public Builder<T> id(Object id) {
+            this.id = id;
+            return this;
+        }
+
+        /**
+         * Sets the (only) value for the header with the specified name.
+         * <p/>
+         * All existing values for the same header will be removed.
+         *
+         * @param name  The name of the header
+         * @param value The value of the header
+         * @return {@code this}
+         */
+        public Builder<T> header(CharSequence name, Object value) {
+            checkNotCookie(name);
```

Comment:
The `checkNotCookie(name)` should be present only when validation is set to true.

We need to come up with a better approach for representing cookies. Going forward, we should build the structured cookie API on top of textual HTTP headers API, making it complementary, instead of enforcing its usage. 

This is important because enforcing structured cookie API restricts the applicability of the API (Think of the WireMock use case). Therefore we need to let API consumers to set/read cookies directly via headers.

For now, make the `checkNotCookie()` somehow optional, but leave any further API improvements to the next increment.

Reply:
This comment also `addHeader(CharSequence, Object)` variant.

Reply:
Hmm... looking at code. This alone is not sufficient. Let me think about this a bit.